import pandas as pd
import numpy as np
import os
import re

configfile: "config/config.yml"

## Load samples & set all columns to strings
df = pd.read_table(config['samples'])

## Check there are accessions
n = len(df.Run)
assert (n > 0), "No accessions were specified. Please add at least one SRA ID"

## Define key values
accessions = df.Run

## Constrain wildcards
wildcard_constraints:
	accession = "(" + '|'.join(accessions) + ")",

# All subsequent paths
raw_path = os.path.join("data", "fastq", "raw")
trim_path = os.path.join("data", "fastq", "trimmed")
fastp_outpath = os.path.join("output", "fastp")
log_path = os.path.join("workflow", "logs")
star_outpath = os.path.join("data", "star")
starlog_path = os.path.join("workflow", "logs", "star")
## unsure how to specify a path outside of this directory, 
## or do I mv/cp the genome and index in here
## or specify in the config?
## or an absolute path in the star.smk input: ? YES thats where it is
get_gtf = os.path.join("/hpcfs/user/a1627307/genome/gencode.v29.annotation.gtf")
fcount_path = os.path.join("output", "counts")



# Define se fastq output only
ALL_FQ = expand(
   os.path.join(raw_path, "{f}_{r}.fastq.gz"),
	r = ['1', '2'], f = accessions
)
ALL_FP = expand(
    os.path.join(fastp_outpath, "{f}_fastp.{suffix}"),
	 f = accessions, suffix = ['json', 'html']
)
ALL_MD5 = expand(
	os.path.join("{path}", "md5sums.txt"),
	path = [raw_path, trim_path]
)
ALL_BAM = expand(
	os.path.join(star_outpath, "{id}", "Aligned.sortedByCoord.out.bam"),
	id = accessions
)
ALL_STARLOGS = expand(
	os.path.join(starlog_path, "{id}_{f}"),
	id = accessions, f = ['Log.final.out', 'SJ.out.tab']
)

ALL_OUTPUTS = []
ALL_OUTPUTS.extend(ALL_MD5)
ALL_OUTPUTS.extend(ALL_FQ)
ALL_OUTPUTS.extend(ALL_FP)
ALL_OUTPUTS.extend(ALL_BAM)
ALL_OUTPUTS.extend(ALL_STARLOGS)

rule all:
	input:
		ALL_OUTPUTS

include: "rules/fasterq-dump.smk"
include: "rules/fastp.smk"
include: "rules/star.smk"
include: "rules/feature-counts.smk"
include: "rules/index-bam.smk"
